### DISCUSSION TITLE
Is overriding custom pythonRequirements at function level allowed?

### DISCUSSION THREAD

**Discussion by 0aprl1** [UnitedIncome/serverless-python-requirements#805] [General]
Hello All

**SETUP**

- python: 3.8.10
- serverless: 1.80.0
- serverless-python-requirements: 5.1.0

**ISSUE**

configuration.yml is a file I use to set ```custom``` properties in serverless.yml, where I define the following noDeploy as a consequence of installing some required library that have unused sub-dependencies:

```
pythonRequirements:
  slim: false
  dockerizePip: true
  usePoetry: false
  noDeploy:
    - boto3
    - botocore
    - docutils
    - jante
    - phonenumbers
    - unidecode
    - sqlmodel
    - sqlalchemy
    - sqlalchemy-stubs
    - psycopg2
    - psycopg2_binary.libs
``` 

I have several lambdas which use the above default pythonRequirements definition. For one particular lambda I have defined, I am attempting the following:

```
test-service:
    name: function-name
    handler: service.handler.handler
    events:
      - http:
          path: /test
          method: get
          authorizer:
            name: authorizer
            resultTtlInSeconds: 3600
          cors: ${self:custom.CORS_CONFIG}
    role: ServiceLambdaExecutionRole
    custom:
      pythonRequirements:
          slim: false
          dockerizePip: true
          usePoetry: false
          noDeploy:
            - boto3
            - botocore
            - docutils
            - jante
            - phonenumbers
            - unidecode
    package: ${self:custom.SERVICE_PACKAGING} 
``` 

Basically for this function, I try to override pythonRequirements at function definition level - the two noDeploy differ in the second not having listed

```
  - sqlmodel
  - sqlalchemy
  - sqlalchemy-stubs
  - psycopg2
  - psycopg2_binary.libs 
```

When I run ```sls package --region eu-west-1 --stage stg```, I am able to package all functions correctly, except for the one defined above ```test-service```, whose zip is excluding everything (11 modules in total) as per the default pythonRequirements, while I would have expected it to exclude only the ones defined at function level (6 in total).

So my question is whether what I am attempting is allowed or not, or if I might have misconfigured something.

Thanks in advance for the support.

**Comment by pgrzesik** [comment]
Hey @0aprl1, `custom` on the function level would be totally ignored, you're also on quite an old version of the Framework, in newer ones it would error out (or at least return a warning). The `noDeploy` option is global and cannot be applied on function level. There's another feature for scoping down dependencies to specific functions which is `modules`. You can read more about them in readme.

**Reply by 0aprl1** [reply]
Thanks for the reply. I have tested, and the issue is that I also have a custom library which is common across all lambdas and I want to remove its subdependencies as well when packaging - so it's no just what is defined directly in the requirements.txt.

I have resorted to adding layers for those lambda that need the removed libraries, and still use noDeploy.
